### PR TITLE
fix(DTFS2-6853): warning text not displayed for supply contract page

### PR DESCRIPTION
**Pull request by vamshidharukef**
## Introduction :pencil2:
On supplier page when user skips some mandatory fields, clicks Next and navigates back to previous page warning text not displayed. User need to go through all the fields which is confusing.

## Resolution :heavy_check_mark:
Added condition to check if status marked as incomplete, when user navigates to previous page warning text should be displayed.

## Miscellaneous :heavy_plus_sign:
Updated tests

### CODE REVIEW

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Please add following tests types:
* Unit tests
* E2E tests

**Review by AlexBramhill — COMMENTED** [review]
I'm going to leave a comment about this ticket on the ticket itself -- I'm not sure I've understood the issue correctly and I think it's best to clarify on the ticket before commenting on here :) 

I'm just posting this review as there's a couple of comments that might be a bit more general ie explaining deep cloning objects

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 1]
```diff
@@ -81,50 +72,44 @@ const hasSubmittedAlwaysShowErrorFields = (allFields, submittedFields) => {
...
 
-  mappedErrors.errorList = filterErrorList(validationErrors.errorList, fields.ALWAYS_SHOW_ERROR_FIELDS);
+  const filteredErrorList = filterErrorList(validationErrors.errorList, fields.ALWAYS_SHOW_ERROR_FIELDS);
 
   return {
-    ...generateErrorSummary(
-      mappedErrors,
-      errorHref,
-    ),
+    ...generateErrorSummary({ ...mappedErrors, errorList: filteredErrorList }, errorHref),
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js]: Changed to previous state

File: portal/server/routes/contract/bond/index.js [hunk 2]
```diff
@@ -91,7 +91,7 @@ const bondDetailsPayloadProperties = [
 ];
 
 const filterBondDetailsPayload = (body) => {
-  const payload = constructPayload(body, bondDetailsPayloadProperties);
+  const payload = constructPayload(body, bondDetailsPayloadProperties,true);
```
Comment (AlexBramhill) [inline portal/server/routes/contract/bond/index.js:94]: same q about allowing empty properties here

File: portal/server/routes/contract/bond/index.js [hunk 3]
```diff
@@ -308,15 +308,15 @@ router.post('/contract/:_id/bond/:bondId/issue-facility', async (req, res) => {
   ];
 
   /**
-   * Add `coverDateConfirmed: true` property to the bond.
+   * Add coverDateConfirmed: true property to the bond.
```
Comment (AlexBramhill) [inline portal/server/routes/contract/bond/index.js:311]: please can we revert this line change

File: portal/server/routes/contract/bond/index.js [hunk 4]
```diff
@@ -445,7 +445,7 @@ router.post('/contract/:_id/bond/:bondId/confirm-requested-cover-start-date', as
   }
 
   /**
-   * Add `coverDateConfirmed: true` property to the bond.
+   * Add coverDateConfirmed: true property to the bond.
```
Comment (AlexBramhill) [inline portal/server/routes/contract/bond/index.js:448]: please can we revert this line change

File: portal/server/routes/contract/about/supplierPage.js [hunk 5]
```diff
@@ -73,7 +83,6 @@ router.get('/contract/:_id/about/supplier', [validateRole({ role: [MAKER] }), pr
     'indemnifier-address-country': mapCountries(countries, indemnifierAddressCountryCode),
     'indemnifier-correspondence-address-country': mapCountries(countries, indemnifierCorrespondenceAddressCountryCode),
   };
-
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/supplierPage.js:76]: could we reinsert these line breaks to improve readability please

File: portal/server/routes/contract/about/supplierPage.js [hunk 6]
```diff
@@ -128,7 +137,7 @@ const supplierSubmissionDetailsFields = [
 router.post('/contract/:_id/about/supplier', provide([INDUSTRY_SECTORS]), async (req, res) => {
   const { industrySectors } = req.apiData;
   const { _id, userToken } = requestParams(req);
-  let submissionDetails = constructPayload(req.body, supplierSubmissionDetailsFields);
+  let submissionDetails = constructPayload(req.body, supplierSubmissionDetailsFields,true);
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/supplierPage.js:140]: same q about allowing empty properties here

File: portal/server/routes/contract/about/supplierPage.js [hunk 7]
```diff
@@ -241,7 +250,6 @@ router.post('/contract/:_id/about/supplier/save-go-back', provide([DEAL, INDUSTR
   }
 
   submissionDetails = industryFields(submissionDetails, industrySectors);
-
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/supplierPage.js:244]: could we please keep line breaks when it improves readability

File: portal/server/routes/contract/about/supplierPage.js [hunk 8]
```diff
@@ -58,7 +56,19 @@ router.get('/contract/:_id/about/supplier', [validateRole({ role: [MAKER] }), pr
     };
     req.session.aboutSupplierFormData = null;
   }
-
+  if(deal.submissionDetails['supplier-companies-house-registration-number']){
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/supplierPage.js:59]: judging by our e2e spec, we do not need to include this here (we already have a passing test for 'A maker picks up a deal in status=Draft, triggers Supplier companies house validation errors'). Either our test is not working and this change is needed (in which case I'd also expect a test rewrite) or this is not needed here?

File: portal/server/routes/contract/about/financialPage.js [hunk 9]
```diff
@@ -55,7 +55,7 @@ const financialSubmissionDetailsProperties = [
 ];
 
 const filterFinancialSubmissionDetailsPayload = (body) => {
-  const payload = constructPayload(body, financialSubmissionDetailsProperties);
+const payload = constructPayload(body, financialSubmissionDetailsProperties,true);
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/financialPage.js:58]: same q about allowing empty properties here

File: portal/server/routes/contract/about/buyerPage.js [hunk 10]
```diff
@@ -66,7 +66,7 @@ const buyerSubmissionDetailsProperties = [
 router.post('/contract/:_id/about/buyer', async (req, res) => {
   const { _id, userToken } = requestParams(req);
 
-  const submissionDetailsPayload = constructPayload(req.body, buyerSubmissionDetailsProperties);
+  const submissionDetailsPayload = constructPayload(req.body, buyerSubmissionDetailsProperties,true);
```
Comment (AlexBramhill) [inline portal/server/routes/contract/about/buyerPage.js:69]: same q about allowing empty properties here

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 11]
```diff
@@ -4,126 +4,115 @@
...
+   return true;
+ }
+
+ return false;
 };
 
 const mapRequiredValidationErrors = (validationErrors, fields) => {
-  const mappedErrors = validationErrors;
-  const allRequiredFields = requiredFieldsArray(fields);
+ const mappedErrors = validationErrors;
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:40]: The formatting across this whole file is off - could you auto format it please

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 12]
```diff
@@ -4,126 +4,115 @@
...
 };
 
 /*
-  ALWAYS_SHOW_ERROR_FIELDS are fields that are:
-  - not required (therefore optional)
-  - BUT if this field has an error, always show it after form submit.
-  - this is the opposite of 'page specific validation rules'
-  - this is currently only used for Companies House validation.
+ ALWAYS_SHOW_ERROR_FIELDS are fields that are:
+ - not required (therefore optional)
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:53]: ... Have you got indent spaces set to 1 by chance :p ?

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 13]
```diff
@@ -4,126 +4,115 @@
...
-
-  mappedErrors.errorList = filterErrorList(validationErrors.errorList, fieldsToReturn);
-
-  return {
-    ...generateErrorSummary(
-      mappedErrors,
-      errorHref,
-    ),
-  };
-};
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:114]: Could we re-add all the deleted line spaces as it makes the code more readable please? ie line spaces between different functions

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js [hunk 14]
```diff
@@ -160,7 +160,7 @@ context('about-supply-contract', () => {
     cy.url().should('eq', relative(`/contract/${dealId}/about/supplier`));
 
     // should only see companies house validation errors
-    partials.errorSummary.errorSummaryLinks().should('have.length', 1);
+    partials.errorSummary.errorSummaryLinks().should('have.length', 12);
```
Comment (AlexBramhill) [inline e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: This shouldnt be changed like this -- the test is checking only that companies house validation errors are present. If there are more than this one error it means that either our test is wrong or our code is wrong

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js [hunk 15]
```diff
@@ -322,4 +322,31 @@ context('about-supply-contract', () => {
...
+    contractAboutSupplier.industrySector().select('Education');
+    contractAboutSupplier.industryClass().select('Cultural education');
+    contractAboutSupplier.smeTypeMicro().click();
+    contractAboutSupplier.supplyContractDescription().type('test');
+    contractAboutSupplier.legallyDistinct().click();
+    contractAboutSupplier.nextPage().click();
+    // re-visit supplier page
+    partials.taskListHeader.itemLink('supplier-and-counter-indemnifier/guarantor').click();
+    // should display missed validation errors
+    partials.errorSummary.errorSummaryLinks().should('have.length', 6);
```
Comment (AlexBramhill) [inline e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:348]: why 6?

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 16]
```diff
@@ -4,126 +4,115 @@
...
+ }
+ const totalFieldValues = Object.keys(fieldValues).filter(
+   (fieldName) =>allFields.includes(fieldName)
+ );
+ if (totalFieldValues.length > 0) {
+   return true;
+ }
+
+ return false;
 };
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:37]: I feel removing: 

```
    && fieldValues[fieldName] !== null
    && fieldValues[fieldName].length > 0
```

is interesting (although I can see why this fixes things) -- are we sure it has no unexpected consiquences?

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 17]
```diff
@@ -4,126 +4,115 @@
...
+
+ mappedErrors.errorList = filterErrorList(validationErrors.errorList,
+fieldsToReturn);
 
+ return {
+   ...generateErrorSummary(mappedErrors, errorHref),
+ };
+};
 const pageSpecificValidationErrors = (validationErrors, fields, submittedFields) => {
   if (validationErrors && validationErrors.errorList) {
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:104]: can we refactor this if statement to be a guard clause please instead to reduce some complexity?

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 18]
```diff
@@ -4,126 +4,115 @@
...
     if (!submittedFields.viewedPreviewPage && hasSubmittedAlwaysShowErrorFields(fields, submittedFields)) {
+      if(submittedFields.status==='Incomplete'){
+        return mapRequiredAndAlwaysShowErrorFields(validationErrors, fields);
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.js:107]: Out of interest why does this fix things please?

File: portal/server/helpers/pageSpecificValidationErrors.test.js [hunk 19]
```diff
@@ -250,6 +250,58 @@ describe('page specific validation errors', () => {
       expect(result).toEqual(expected);
     });
 
+    it('displays validation errors when user skips mandatory fields', () => {
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.test.js:253]: formatting

also companies house is optional isnt it?

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 20]
```diff
@@ -4,126 +4,115 @@
...
-    return true;
-  }
-
-  return false;
+ const allFields = allFieldsArray(fields);
+
+ if (fieldValues.viewedPreviewPage) {
+   return true;
+ }
+ const totalFieldValues = Object.keys(fieldValues).filter(
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/server/helpers/pageSpecificValidationErrors.js:29]: We should add unit tests around this area checking and documenting when fieldValues is not null, when it is null. When .length is greater than 0 and when .length is 0 -- this is towards our goals of 100% (useful) test coverage

File: portal/server/helpers/pageSpecificValidationErrors.test.js [hunk 21]
```diff
@@ -250,6 +250,58 @@ describe('page specific validation errors', () => {
       expect(result).toEqual(expected);
     });
```
Comment (AlexBramhill) [inline portal/server/helpers/pageSpecificValidationErrors.test.js:252]: Can we add tests around the fix please

File: portal/server/routes/contract/loan/index.js [hunk 22]
```diff
@@ -94,7 +94,7 @@ const loanGuaranteeDetailsPayloadProperties = [
 ];
 
 const filterLoanGuaranteeDetailsPayload = (body) => {
-  const payload = constructPayload(body, loanGuaranteeDetailsPayloadProperties);
+const payload = constructPayload(body, loanGuaranteeDetailsPayloadProperties,true);
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/routes/contract/loan/index.js:97]: It will insert the empty properties but when we insert the value, the property will update and code will check the validation also for all the pages.

File: portal/server/routes/contract/loan/index.js [hunk 23]
```diff
@@ -300,15 +300,15 @@ router.post('/contract/:_id/loan/:loanId/issue-facility', async (req, res) => {
   ];
 
   /**
-   * Add `coverDateConfirmed: true` property to the bond.
+   * Add coverDateConfirmed: true property to the bond.
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/routes/contract/loan/index.js:303]: Done

File: portal/server/routes/contract/loan/index.js [hunk 24]
```diff
@@ -431,7 +431,7 @@ router.post('/contract/:_id/loan/:loanId/confirm-requested-cover-start-date', pr
   }
 
   /**
-   * Add `coverDateConfirmed: true` property to the bond.
+   * Add coverDateConfirmed: true property to the bond.
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/routes/contract/loan/index.js:434]: Done

File: portal/server/routes/contract/bond/index.js (hunk 2 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:94]: It will insert the empty properties but when we insert the value, the property will update and code will check the validation also for all the pages.

File: portal/server/routes/contract/bond/index.js (hunk 3 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:311]: Done

File: portal/server/routes/contract/bond/index.js (hunk 4 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:448]: Done

File: portal/server/routes/contract/about/supplierPage.js (hunk 6 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/supplierPage.js:140]: It will insert the empty properties but when we insert the value, the property will update and code will check the validation also for all the pages.

File: portal/server/routes/contract/about/supplierPage.js (hunk 5 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/supplierPage.js:76]: Done

File: portal/server/routes/contract/about/supplierPage.js (hunk 7 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/supplierPage.js:244]: Done

File: portal/server/routes/contract/about/financialPage.js (hunk 9 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/financialPage.js:58]: It will insert the empty properties but when we insert the value, the property will update and code will check the validation also for all the pages.

File: portal/server/routes/contract/about/buyerPage.js (hunk 10 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/buyerPage.js:69]: It will insert the empty properties but when we insert the value, the property will update and code will check the validation also for all the pages.

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 11 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:40]: Done

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 12 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:53]: Done

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 13 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:114]: Done

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 14 above)
Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: When we entered the companies and come back second time the same page. Still we need to show the error message. So we update the code, that is reason we update the e2e tests also.

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 15 above)
Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:348]: It returns 6 summary links hence comparing with 6

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 25]
```diff
@@ -4,126 +4,115 @@
...
-    ...generateErrorSummary(
-      mappedErrors,
-      errorHref,
-    ),
-  };
+ return {
+   ...generateErrorSummary({ ...mappedErrors, errorList:
+filteredErrorList }, errorHref),
+ };
 };
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:86]: Done

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 18 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:107]: If supplier page is not completed means status in incomplete. It will check the status and display the error(required field and optional field error)

File: portal/server/helpers/pageSpecificValidationErrors.test.js (hunk 19 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.test.js:253]: Done, yes

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 26]
```diff
@@ -4,126 +4,115 @@
...
 
 const mapAlwaysShowErrorFields = (validationErrors, fields) => {
-  const mappedErrors = validationErrors;
+ const mappedErrors = { ...validationErrors };
 
-  mappedErrors.errorList = filterErrorList(validationErrors.errorList, fields.ALWAYS_SHOW_ERROR_FIELDS);
+ const filteredErrorList =
+filterErrorList(validationErrors.errorList,
+fields.ALWAYS_SHOW_ERROR_FIELDS);
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:81]: Done

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 17 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:104]: Done

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:29]: We have test for scenarios `should return true when field values length is greater than 0`, `should return true when there is a fields.viewedPreviewPage is true`, `should return false when field values length is NOT greater than 0`

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/server/helpers/pageSpecificValidationErrors.js:77]: Sorry the above is removing line 72 `  const mappedErrors = { ...validationErrors };` as it's not needed in the suggested imp -- redues the amount of cognative complexity

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 14 above)
Comment (AlexBramhill) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: The comment says `    // should only see companies house validation errors` -- It feels that as a user if I search for a companies house number, I would not expect the rest of the page to error at unfilled in data (especially as it's at the top of the page!). @abhi-markan do you know about any intended behaviour in this regard or should Vamshi contact the ui/ux guys please?

File: portal/server/routes/contract/loan/index.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/routes/contract/loan/index.js:97]: ConstructPayload function can expect canPropertyBeEmpty variable. Default the variable is false instead of false we passed true. So constructPayload function will construct the payload variables with empty value. That payload will save in DB object. Second time we come to the same page we fetch all the details from DB for the page and check the value is exist or not. If value is null then will display the error otherwise no error will display. When we enter the value for the empty object, the post will construct new payload and update the DB.

File: portal/server/routes/contract/bond/index.js (hunk 2 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:94]: ConstructPayload function can expect canPropertyBeEmpty variable. Default the variable is false instead of false we passed true. So constructPayload function will construct the payload variables with empty value. That payload will save in DB object. Second time we come to the same page we fetch all the details from DB for the page and check the value is exist or not. If value is null then will display the error otherwise no error will display. When we enter the value for the empty object, the post will construct new payload and update the DB.

File: portal/server/routes/contract/about/financialPage.js (hunk 9 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/financialPage.js:58]: ConstructPayload function can expect canPropertyBeEmpty variable. Default the variable is false instead of false we passed true. So constructPayload function will construct the payload variables with empty value. That payload will save in DB object. Second time we come to the same page we fetch all the details from DB for the page and check the value is exist or not. If value is null then will display the error otherwise no error will display. When we enter the value for the empty object, the post will construct new payload and update the DB.

File: portal/server/routes/contract/about/buyerPage.js (hunk 10 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/about/buyerPage.js:69]: ConstructPayload function can expect canPropertyBeEmpty variable. Default the variable is false instead of false we passed true. So constructPayload function will construct the payload variables with empty value. That payload will save in DB object. Second time we come to the same page we fetch all the details from DB for the page and check the value is exist or not. If value is null then will display the error otherwise no error will display. When we enter the value for the empty object, the post will construct new payload and update the DB.

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 14 above)
Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: When the user searches companies house number with invalid data or empty filed just the companies house validation error is displayed. If valid companies house number is entered that values are populated for rest of of the fields except for supplier type, supplier correspondence address, industry sector, industry class, sme type, description, guarantor and if user clicks on Next and navigates back validation errors displayed only for the missing fields.

Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: Had a call with Abhi on this. Abhi confirmed that the error displayed is as expected.

Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: I checked the test the companies house validation error is added to the list of errors on supplier page so ether the comment needs to be removed or should we need to come up with new test.

Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: I came up with below code and the test pass which checks length of 1 for companies house. Please let me know if this is correct approach to test companies house validation error.                                                                          ```
it('trigger just Supplier companies house validation error', () => {
    cy.login(BANK1_MAKER1);
    // dashboard.createNewSubmission().click();
    cy.visit('/dashboard');
    // Select "Create new" and choose "Bond Support" as the scheme type
    cy.get('[data-cy="CreateNewSubmission"]').click();
    cy.get('[data-cy="scheme-bss"]').click();
    cy.get('[data-cy="continue-button"]').click();
    cy.get('[data-cy="criteriaMet-true"]').click();
    cy.get('[data-cy="submit-button"]').click();
    cy.get('[data-cy="bankInternalRefName"]').type('TestBank1903');
    cy.get('[data-cy="additionalRefName"]').type('TestBank1903');
    cy.get('[data-cy="submit-button"]').click();

    // Click on "View Details" under the "About the Supplier" contract
    cy.get('[data-cy="ViewAboutSupplierDetails"]').click();
    // Enter an invalid Companies House registration number
    contractAboutSupplier.supplierCompaniesHouseRegistrationNumber().type('TEST');

    // Click on Search
    contractAboutSupplier.supplierSearchCompaniesHouse().click();

    // Check that the correct validation error is displayed
    partials.errorSummary.errorSummaryLinks().should('have.length', 1);
    contractAboutSupplier.expectError('Enter a valid Companies House registration number');
    // Enter an empty Companies House registration number
    contractAboutSupplier.supplierCompaniesHouseRegistrationNumber().clear();
    contractAboutSupplier.supplierSearchCompaniesHouse().click();
    // Check that the correct validation error is displayed
    partials.errorSummary.errorSummaryLinks().should('have.length', 1);
    contractAboutSupplier.expectError('Enter a valid Companies House registration number'); 
});
```

File: portal/server/routes/contract/about/buyerPage.js (hunk 10 above)
Comment (AlexBramhill) [reply portal/server/routes/contract/about/buyerPage.js:69]: I get you. It feels really odd this is needed now. Will close other threads so this ones the only one on the go about it. I'm going to do a bit of testing tomorrow because there's some odd behaviour going on in the existing component on `main` which i feel ties into this

File: portal/server/routes/contract/bond/index.js (hunk 4 above)
Comment (AlexBramhill) [reply portal/server/routes/contract/bond/index.js:448]: might be an issue with my git but im not seeing it

File: portal/server/routes/contract/bond/index.js (hunk 3 above)
Comment (AlexBramhill) [reply portal/server/routes/contract/bond/index.js:311]: might be an issue with my git but im not seeing it

File: portal/server/routes/contract/loan/index.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/server/routes/contract/loan/index.js:434]: might be an issue with my git but im not seeing it

**Review by AlexBramhill — COMMENTED** [review]
very close, I am still trying to understand this bug completely, but looking good. Coudl we doulbe check we have tests for 
1. not entering a companies house number and check errors (without navigating off page)
2. going next page and back to this page with a few specific fields missing
3. going to error summary with certain fields missing

I hope we already have these

File: portal/server/helpers/pageSpecificValidationErrors.js [hunk 27]
```diff
@@ -81,49 +69,50 @@ const hasSubmittedAlwaysShowErrorFields = (allFields, submittedFields) => {
...
   return {
-    ...generateErrorSummary(
-      mappedErrors,
-      errorHref,
-    ),
+    ...generateErrorSummary(mappedErrors, errorHref),
   };
 };
-
 const pageSpecificValidationErrors = (validationErrors, fields, submittedFields) => {
```
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:103]: Added new line

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 15 above)
Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:348]: Added comment

File: e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js (hunk 14 above)
Comment (vamshidharukef) [reply e2e-tests/portal/cypress/e2e/journeys/maker-about-supply-contract/fill-in-about-supply-contract-validation.spec.js:163]: Added test to check just companies house validation error

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:77]: Not sure what you are referring to.

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 17 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.js:104]: Removed guard class

File: portal/server/helpers/pageSpecificValidationErrors.test.js (hunk 21 above)
Comment (vamshidharukef) [reply portal/server/helpers/pageSpecificValidationErrors.test.js:218]: Added test to check if field values are null

File: portal/server/routes/contract/bond/index.js (hunk 3 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:311]: This is done. I cannot see at my end.

File: portal/server/routes/contract/bond/index.js (hunk 4 above)
Comment (vamshidharukef) [reply portal/server/routes/contract/bond/index.js:448]: This is done. I cannot see at my end.

File: portal/server/routes/contract/loan/index.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (vamshidharukef) [reply portal/server/routes/contract/loan/index.js:434]: This is done. I cannot see at my end.

File: portal/server/helpers/pageSpecificValidationErrors.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/server/helpers/pageSpecificValidationErrors.js:83]: Doesnt matter -- happy to leave for now to get this merged. Can address this refactor in another ticket

**Review by AlexBramhill — APPROVED** [review]
great job!

**Review by abhi-markan — APPROVED** [review]
